### PR TITLE
Fix diagnostics for async block cloning

### LINKED ISSUE #121547: Compiler Suggests Cloning an `async` Block
### Code

```Rust
// https://github.com/HTGAzureX1212/HarTex/blob/491fe72da9042d5a5c7f61d0d5cdaf6e9b112057/discord-frontend/hartex-discord-leader/src/kafka.rs#L79-L120
```

### Current output

```Shell
error[E0382]: use of moved value: `producer`
   --> hartex-discord-leader\src\kafka.rs:78:23
    |
68  |   async fn inbound<'a, Q>(shards: Vec<&'a mut Shard<Q>>, producer: FutureProducer) -> miette::Result<()>
    |                                                          -------- move occurs because `producer` has type `FutureProducer`, which does not implement the `Copy` trait
...
74  |       loop {
    |       ---- inside of this loop
...
77  |           for shard in shards {
    |           ------------------- inside of this loop
78  |               set.spawn(async move {
    |  _______________________^
79  | |                 while let Some(result) = shard.next().await {
80  | |                     match result {
81  | |                         Ok(message) => {
...   |
95  | |                             if let Err((error, _)) = producer
    | |                                                      -------- use occurs due to use in coroutine
...   |
119 | |                 }
120 | |             });
    | |_____________^ value moved here, in previous iteration of loop
    |
help: consider cloning the value if the performance cost is acceptable
    |
120 |             }.clone());
    |              ++++++++
```

### Desired output

```Shell
error[E0382]: use of moved value: `producer`
   --> hartex-discord-leader\src\kafka.rs:78:23
    |
68  |   async fn inbound<'a, Q>(shards: Vec<&'a mut Shard<Q>>, producer: FutureProducer) -> miette::Result<()>
    |                                                          -------- move occurs because `producer` has type `FutureProducer`, which does not implement the `Copy` trait
...
74  |       loop {
    |       ---- inside of this loop
...
77  |           for shard in shards {
    |           ------------------- inside of this loop
78  |               set.spawn(async move {
    |  _______________________^
79  | |                 while let Some(result) = shard.next().await {
80  | |                     match result {
81  | |                         Ok(message) => {
...   |
95  | |                             if let Err((error, _)) = producer
    | |                                                      -------- use occurs due to use in coroutine
...   |
119 | |                 }
120 | |             });
    | |_____________^ value moved here, in previous iteration of loop
    |
help: consider cloning the value if the performance cost is acceptable
    |
95  |                               if let Err((error, _)) = producer.clone()
    |                                                                ++++++++
```

### Rationale and extra context

I expect that the `.clone()` invocation to be added for the value that is moved, but for some reason it is doing it for the entire `async` closure.

### Other cases

_No response_

### Rust Version

```Shell
rustc 1.78.0-nightly (8f359beca 2024-02-23)
binary: rustc
commit-hash: 8f359beca4e58bc3ae795a666301a8f47023044c
commit-date: 2024-02-23
host: x86_64-pc-windows-msvc
release: 1.78.0-nightly
LLVM version: 18.1.0
```

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by wutchzone**
Closes #121547

r? diagnostics

### CODE REVIEW

File: compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs [hunk 1]
```diff
@@ -463,6 +464,15 @@ impl<'cx, 'tcx> MirBorrowckCtxt<'cx, 'tcx> {
                 } else if let UseSpans::FnSelfUse { kind: CallKind::Normal { .. }, .. } = move_spans
                 {
                     // We already suggest cloning for these cases in `explain_captures`.
+                } else if let UseSpans::ClosureUse {
+                    closure_kind:
+                        ClosureKind::Coroutine(CoroutineKind::Desugared(_, CoroutineSource::Block)),
+                    args_span: _,
+                    capture_kind_span: _,
+                    path_span,
```
Comment (wutchzone) [inline compiler/rustc_borrowck/src/diagnostics/conflict_errors.rs:472]: I am not sure what is more fitting, whether `path_span` or `capture_kind_span`. Both seem to be working correctly with the provided test.

File: tests/ui/borrowck/cloning-in-async-block-121547.stderr [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+LL |       for _ in 0..10 {
+   |       -------------- inside of this loop
+LL | /         async {
+LL | |             drop(value);
+   | |                  ----- use occurs due to use in coroutine
+LL | |
+LL | |         }.await
+   | |_________^ value moved here, in previous iteration of loop
+   |
+help: consider cloning the value if the performance cost is acceptable
```
Comment (compiler-errors) [inline tests/ui/borrowck/cloning-in-async-block-121547.stderr:15]: I feel like this error message is not *right* in spirit. Really we should be saying something like "cloning the value [if this is the behavior that you intended]".

Cloning here could have different semantics than actually moving the value into the async block, for example.

**Review by compiler-errors — COMMENTED** [review]
This seems fine, but I kind of wish that the message noted that the semantics change due to the clone.

File: tests/ui/borrowck/cloning-in-async-block-121547.stderr (hunk 2 above)
Comment (wutchzone) [reply tests/ui/borrowck/cloning-in-async-block-121547.stderr:15]: Just to make sure, are you suggesting we should change the help message in general for cloning or only for this particular case in the async block?

Comment (compiler-errors) [reply tests/ui/borrowck/cloning-in-async-block-121547.stderr:15]: I don't actually know :/ I guess this is a more general problem than what this aims to fix

**Review by compiler-errors — COMMENTED** [review]
Does this work for cases like:

```
let x = async {
    drop(value);
}.await;
use(value);
```

Outside of a loop?

**Comment by wutchzone** [comment]
FYI: due to the #121652, the suggestion is even more funnier :grinning: 

```text
error[E0382]: use of moved value: `value`
 --> src/main.rs:3:9
  |
1 |   async fn clone_async_block(value: String) {
  |                              ----- move occurs because `value` has type `String`, which does not implement the `Copy` trait
2 |       for _ in 0..10 {
  |       -------------- inside of this loop
3 | /         async { //~ ERROR: use of moved value: `value` [E0382]
4 | |             drop(value);
  | |                  ----- use occurs due to use in coroutine
5 | |             //~^ HELP: consider cloning the value if the performance cost is acceptable
6 | |         }.await
  | |_________^ value moved here, in previous iteration of loop
  |
help: consider moving the expression out of the loop so it is only moved once
  |
2 ~     let mut value = await;
3 ~     for _ in 0..10 {
4 |         async { //~ ERROR: use of moved value: `value` [E0382]
5 |             drop(value);
6 |             //~^ HELP: consider cloning the value if the performance cost is acceptable
7 ~         }.value
  |
```

**Comment by wutchzone** [comment]
> Does this work for cases like:
> ```rust
> let x = async {
>    drop(value);
> }.await;
> use(value);
> ```
> Outside of a loop?

nightly:
```
error[E0382]: use of moved value: `value`
 --> src/main.rs:8:13
  |
4 |   async fn clone_async_block(value: String) {
  |                              ----- move occurs because `value` has type `String`, which does not implement the `Copy` trait
5 |       let _x = async {
  |  ______________-
6 | |         drop(value);
  | |              ----- variable moved due to use in coroutine
7 | |     }.await;
  | |_____- value moved here
8 |       consume(value);
  |               ^^^^^ value used here after move
  |
help: consider cloning the value if the performance cost is acceptable
  |
7 |     }.clone().await;
  |      ++++++++

```

after my patch:
```
error[E0382]: use of moved value: `value`
 --> src/main.rs:8:13
  |
4 |   async fn clone_async_block(value: String) {
  |                              ----- move occurs because `value` has type `String`, which does not implement the `Copy` trait
5 |       let _x = async {
  |  ______________-
6 | |         drop(value);
  | |              ----- variable moved due to use in coroutine
7 | |     }.await;
  | |_____- value moved here
8 |       consume(value);
  |               ^^^^^ value used here after move
  |
help: consider cloning the value if the performance cost is acceptable
  |
6 |         drop(value.clone());
  |                   ++++++++
```

I guess it works even for this case.

File: tests/ui/borrowck/cloning-in-async-block-121547.stderr (hunk 2 above)
Comment (wutchzone) [reply tests/ui/borrowck/cloning-in-async-block-121547.stderr:15]: May I suggest creating an issue where we can talk about the appropriate wording? I would like to also share my opinion, however, I think this is not the right PR for that.

**Review by compiler-errors — APPROVED** [review]

**Pull request merged**